kame/kame/icmp6dump/icmp6dump.8
.\"	$KAME: icmp6dump.8,v 1.3 2000/11/09 03:04:08 itojun Exp $
.\"
.\" Copyright (C) 1995, 1996, 1997, and 1998 WIDE Project.
.\" All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the project nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd May 17, 1998
.Dt ICMP6DUMP 8
.Os KAME
.Sh NAME
.Nm icmp6dump
.Nd dump ICMPv6 packets on the network
.Sh SYNOPSIS
.Nm
.Op Ar type...
.Sh DESCRIPTION
.Nm Icmp6dump
catches ICMPv6 packets toward the host, and displays the content of packets.
Packet can be filtered using the optional argument
.Ar type .
.Ar Type
must be decimal integer value for the ICMPv6 type.
.Sh RETURN VALUES
The program exits with 0 on success, non-zero on failures.
.\" .Sh SEE ALSO
.\" (to be written)
.Sh HISTORY
The
.Nm
command first appeared in WIDE Hydrangea IPv6 protocol stack kit.
.\" .Sh BUGS
.\" (to be written)


CHANGELOG
$KAME: CHANGELOG,v 1.2458 2003/08/18 05:39:09 itojun Exp $

<200308>
Sun Aug 17 JST 2003  itojun@iijlab.net
	* sys/netinet{,6}/in{,6}_proto.c: enforce inbound IPsec policy on
	  protocols that falls into raw socket as a wildcard case.

2003-08-15  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/kame/pim6sd/{cfparse.y, mrt.[ch], pim6_proto.c, rp.c}
	Now static-RP configuration can work together with bootstrap-
	message-based RP configuration

2003-08-10  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/{freebsd5, kame}
	sync with FreeBSD-5.1 RELEASE (behaviour is not confirmed yet)

2003-08-07  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* kame/kame/had/mpa.c,haadisc.c
	A home address should be passed in the source address of a mobile
	solicit pakcet. This problem was reported and donated the code by
	Kenichi Yajima <yajima@netlab.nec.co.jp>
	
2003-08-07  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/cftoken.l: use a stronger lexical check for
	interface names.
	From: Hajimu UMEMOTO <ume@mahoroba.org>
	* kame/kame/dhcp6/config.c (configure_interface): check interfaces
	specified in the configuration file really exist.

2003-08-05  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* removed vrrp6 support because it might be a patent infringement.
	  http://www.ietf.org/ietf/IPR/cisco-ipr-draft-ietf-vrrp-ipv6-spec.txt

2003-08-01  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: implemented various clarifications and some new
	  features:
	- supported DNS name and NTP servers options
	- dhcp6s now uses the server/relay port to send a relay-rely
	  message
	- dhcp6s now ignores a rebind message when it cannot find a
	  binding, according to prefix-delegation-04
	- dhcp6relay now accepts relay-reply messages on the socket
	  listening on the server/relay port
	* The DHCPv6 implementation is now built by default under
	  *bsd/usr.sbin/.

<200307>
Thu Jul 31 19:19:48 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c,mip6_cncore.c
	sending/receiving a binding refresh request message is supported.

Tue Jul 29 17:09:15 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.[hc]
	select a CoA using the source address selection like algorithm.
	suggested by Francis.Dupont@enst-bretagne.fr.

2003-07-28  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* sys/netinet6/{ip6_output.c, ip6_var.h, mip6_cncore.c,
	mip6_hacore.[ch], nd6.c}:
	CN or HA dosen't add a type2 routing header to a Binding
	Acknowledgement packet when a deregistration packet was issued
	on home link and it was failed by some reasons.
	reported by Yukiyo.Akisada@jp.yokogawa.com and v6pc cert.
	team.

Mon Jul 28 20:03:46 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_cncore.c,mip6_mncore.[hc],in6.c,nd6_rtr.c
	  kame/sys/net/if_hif.[hc]
	assign a CoA separately for each home interface, since a home
	interface is logically independent from each other.

Mon Jul 28 14:38:11 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_hacore.c,mip6_cncore.c:
	- fixed a bug not sending a binding ack when DAD for a link-local
	  address is failed.  (en-bugged during re-structuring BC logic.)
	- fixed a bug using RTHDR2 with Mobility Headers other than
	  a binding ack.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

Fri Jul 25 19:12:15 JST 2003  itojun@iijlab.net
	* sys/netinet/ah_core.c: hmac-ripemd160 support

Fri Jul 25 18:07:43 JST 2003 sakane@kame.net
	* kame/sys/netkey/key.c
	fixed that the kernel crashed when key_spdacquire() was called
	because key_spdacquire() had been implemented imcopletely.

Thu Jul 24 16:11:06 JST 2003	keiichi@iij.ad.jp
	* mip6 related files.
	simplify the prefix and advertising router list management
	mechanism on a mobile node.

2003-07-20  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/config.c (configure_duid): fixed a bug of
	allocating short memory.  Upgrading is required if you configure
	dhcp6s with the duid statement.

Sat Jul 19 18:12:01 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_aesctr.c: support draft-ietf-ipsec-ciph-aes-ctr-03
	  as there's no official DOI assignment yet, we use private DOI number.
	* sys/netinet6/ah_aesxcbcmac.c: support
	  draft-ietf-ipsec-ciph-aes-xcbc-mac-03.
	  as there's no official DOI assignment yet, we use private DOI number.

Tue Jul 15 20:02:05 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_rijndael.c: simplify and update rijndael code.
	  markus@openbsd

2003-07-14  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: revised the relay agent implementation
	based on dhcpv6-28.
	- dhcp6relay now relays packets between clients and servers using
	  relay forward/reply messages.
	- dhcp6s now accepts relay forward messages and replies with relay
	  reply messages.

Sat Jul 12 15:05:23 JST 2003	suz@crl.hitachi.co.jp
	* openbsd/sys/netinet/igmp_var.h
	makes IGMPV3-kernel compilable on openbsd.

Thu Jul 10 21:35:04 JST 2003	keiichi@iij.ad.jp
	* kame/kame/had/halist.c
	fixed a bug that a DHAAD reply message only includes one global
	address per home agent.  the message must include all global
	addresses of all home agents.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

2003/07/10 21:07:50 JST kjc@csl.sony.co.jp
	the first step for transition to pf/altq.
	marge pf/altq into the existing altq-3.
	allow both pf/altq and altq-3 to coexit.
	currently, ALTQ3_COMPAT and ALTQ3_CLFIER_COMPAT are defined
	in sys/altq/altq.h to support altq-3.
	altq-3 will remain in kame for research experiments but will
	not be merged into bsd releases.

2003-07-08  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* kame/sys/netinet6/{in6_ifattach.c,ip6_mroute.c,ip6_mroute.h}
	protect against interface removal.

Tue Jul  8 18:51:06 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c
	insert alternate careof address sub-option when creating a
	binding update for the home registration.  This sub-option is
	MUST when we protect the message with ESP and strictly speaking,

kame/kame/icmp6dump/icmp6dump.c
/*
 * Copyright (C) 1995, 1996, 1997, and 1998 WIDE Project.
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the project nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/param.h>
#include <sys/socket.h>

#include <net/if.h>
#include <netinet/in.h>
#include <netinet/ip6.h>
#include <netinet/icmp6.h>
#include <unistd.h>
#include <stdio.h>
#include <err.h>
#include <errno.h>
#include <string.h>
#include <stdlib.h>
#include <arpa/inet.h>

int sock;
void sock_open(void);
void dump(void);

int
main(argc, argv)
	int argc;
	char *argv[];
{
	fd_set *fdsetp;
	size_t nfdset;
	int i;
	struct icmp6_filter filt;
	
	sock_open();

#ifdef ICMP6_FILTER
	if (argc == 1)
		ICMP6_FILTER_SETPASSALL(&filt);
	else {
		int type;

		ICMP6_FILTER_SETBLOCKALL(&filt);
		argc--; argv++;
		while (argc-- > 0) {
			type = atoi(*argv);
			ICMP6_FILTER_SETPASS(type, &filt);
			argv++;
		}
	}
	if (setsockopt(sock, IPPROTO_ICMPV6, ICMP6_FILTER, &filt,
			sizeof(filt)) < 0) {
		perror("setsockopt(ICMP6_FILTER)");
		exit(-1);
	}
#endif /*ICMP6_FILTER*/

	nfdset = howmany(sock, NFDBITS);
	if ((fdsetp = malloc(nfdset)) == NULL)
		err(1, "malloc");
	memset(fdsetp, 0, nfdset);
	for (;;) {
		FD_SET(sock, fdsetp);
		if ((i = select(sock + 1, fdsetp, NULL, NULL, NULL)) < 0)
			perror("select");
		if(i == 0)
			continue;
		else
			dump();
	}
	exit(0);
}

void
dump()
{
	int i, j;
	struct icmp6_hdr *icmp6;
	u_char buf[1024];
	struct sockaddr_in6 from;
	int from_len = sizeof(from);
	char ntop_buf[256];

	if ((i = recvfrom(sock, buf, sizeof(buf), 0,
			  (struct sockaddr *)&from,
			  &from_len)) < 0)
		return;

	if (i < sizeof(struct icmp6_hdr)) {
		printf("too short!\n");
		return;
	}
	icmp6 = (struct icmp6_hdr *)buf;

	printf("from %s, ", inet_ntop(AF_INET6, &from.sin6_addr,
				      ntop_buf, sizeof(ntop_buf)));
	switch (icmp6->icmp6_type) {
	case ICMP6_DST_UNREACH:
		printf("type=unreach,");
		switch(icmp6->icmp6_code) {
		case ICMP6_DST_UNREACH_NOROUTE:
			printf("code=no route\n");
			break;
		case ICMP6_DST_UNREACH_ADMIN:
			printf("code=admin\n");
			break;
		case ICMP6_DST_UNREACH_BEYONDSCOPE:
			printf("code=beyond scope\n");
			break;
		case ICMP6_DST_UNREACH_ADDR:
			printf("code=address\n");
			break;
		case ICMP6_DST_UNREACH_NOPORT:
			printf("code=port\n");
			break;
		}
		break;
	case ICMP6_PACKET_TOO_BIG:
		printf("type=packet too big\n");
		break;
	case ICMP6_TIME_EXCEEDED:
		printf("type=time exceed,");
		switch(icmp6->icmp6_code) {
		case ICMP6_TIME_EXCEED_TRANSIT:
			printf("code=in trans\n");
			break;
		case ICMP6_TIME_EXCEED_REASSEMBLY:
			printf("code=in reass\n");
			break;
		}
		break;
	case ICMP6_PARAM_PROB:
		printf("type=parameter problem,");
		switch(icmp6->icmp6_code) {
		case ICMP6_PARAMPROB_HEADER:
			printf("code=header\n");
			break;
		case ICMP6_PARAMPROB_NEXTHEADER:
			printf("code=next header\n");
			break;
		case ICMP6_PARAMPROB_OPTION:
			printf("code=option\n");
			break;
		}
		break;
	case ICMP6_ECHO_REQUEST:
		printf("type=icmp echo request\n");
		break;
	case ICMP6_ECHO_REPLY:
		printf("type=icmp echo reply\n");
		break;
	case ICMP6_FQDN_QUERY:
		printf("type=icmp FQDN query\n");
		break;
	case ICMP6_FQDN_REPLY:
		printf("type=icmp FQDN reply\n");
		break;
	case ICMP6_MEMBERSHIP_QUERY:
		printf("type=group query\n");
		break;
	case ICMP6_MEMBERSHIP_REPORT:
		printf("type=group report\n");
		break;
	case ICMP6_MEMBERSHIP_REDUCTION:
		printf("type=group termination\n");
		break;
	case ND_ROUTER_SOLICIT:
		printf("type=router solicitation\n");
		break;
	case ND_ROUTER_ADVERT:
		printf("type=router advertisement\n");
		break;
	case ND_NEIGHBOR_SOLICIT:
		printf("type=neighbor solicitation\n");
		break;
	case ND_NEIGHBOR_ADVERT:
		printf("type=neighbor advertisement\n");
		break;
	case ND_REDIRECT:
		printf("type=redirect\n");
		break;
	default:
		printf("type=unknown\n");
		break;
	}

	j = 0;
	printf("  ");
	while (j < i) {
		if (buf[j] <= 0xf)
			putchar('0');
		printf("%x", buf[j++]);
		if (j % 2 == 0)
			putchar(' ');
		if (j % 16 == 0)
			printf("\n  ");
	}
	printf("\n");
	if (j % 16 != 0)
		printf("\n");
	fflush(stdout);
}


void
sock_open()
{
	struct sockaddr_in6 me;
	
	if ((sock = socket(AF_INET6, SOCK_RAW, IPPROTO_ICMPV6)) < 0)
		err(1, "socket");

	memset(&me, 0, sizeof(struct sockaddr_in6));
	me.sin6_len = sizeof(struct sockaddr_in6);
	me.sin6_family = AF_INET6;

	if (bind(sock, (struct sockaddr *)&me, sizeof(me)) < 0)
		err(1, "bind");
}
